File: memory_management.cpp
#include <iostream>

int main() {
    // Dynamically allocating memory for an array on the heap
    int* arr = new int[5];

    // Assigning values to the array
    for (int i = 0; i < 5; i++) {
        arr[i] = i + 1;
    }

    // Freeing memory manually
    delete[] arr;

    // After this point, arr is dangling if accessed again.
    // Uncommenting the following line will cause undefined behavior.
    // std::cout << arr[0] << std::endl; // Dangling pointer error

    std::cout << "Memory manually freed with delete." << std::endl;

    return 0;
}

File: memory_management.rs
fn main() {
    // Dynamically allocating memory on the heap
    let vec = Box::new(vec![1, 2, 3, 4, 5]);

    // Borrowing: `vec` is borrowed by `borrowed_vec`
    let borrowed_vec = &vec;

    // Ownership: Ownership is transferred to `new_vec`
    let new_vec = vec;

    // println!("{:?}", borrowed_vec); // This would cause a compile-time error because `vec`'s ownership is moved to `new_vec`
    
    println!("{:?}", new_vec);
}
